File: mobile/lib/core/network/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/env_config.dart';

class ApiClient {
  final String baseUrl;
  static String? _accessToken;
  static String? _refreshToken;

  ApiClient({String? baseUrl}) : baseUrl = baseUrl ?? EnvConfig.apiBaseUrl;

  static void setTokens(String access, [String? refresh]) {
    _accessToken = access;
    _refreshToken = refresh;
  }

  static void clearTokens() {
    _accessToken = null;
    _refreshToken = null;
  }

  static bool get isAuthenticated => _accessToken != null;

  Map<String, String> _getHeaders() {
    final headers = <String, String>{'Content-Type': 'application/json'};
    if (_accessToken != null) {
      headers['Authorization'] = 'Bearer $_accessToken';
    }
    return headers;
  }

  Future<Map<String, dynamic>> checkHealth() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/health'),
        headers: _getHeaders(),
      );
      if (response.statusCode == 200) {
        return json.decode(response.body) as Map<String, dynamic>;
      }
      return {'status': 'unhealthy', 'message': 'HTTP ${response.statusCode}'};
    } catch (e) {
      return {'status': 'error', 'message': e.toString()};
    }
  }

  // ------------------ AUTH ------------------
  Future<Map<String, dynamic>> register(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/auth/register'),
      headers: _getHeaders(),
      body: json.encode(data),
    );
    final resBody = json.decode(response.body);
    if (response.statusCode != 201) {
      throw Exception(resBody['error']?['message'] ?? 'Registration failed');
    }
    return resBody;
  }

  Future<Map<String, dynamic>> sendOtp(String target, [String otpType = 'REGISTRATION']) async {
    final response = await http.post(
      Uri.parse('$baseUrl/auth/send-otp'),
      headers: _getHeaders(),
      body: json.encode({'target': target, 'otp_type': otpType}),
    );
    final resBody = json.decode(response.body);
    if (response.statusCode != 200) {
      throw Exception(resBody['error']?['message'] ?? 'Failed to send OTP');
    }
    return resBody;
  }

  Future<Map<String, dynamic>> verifyOtp(String target, String otpCode, [String otpType = 'REGISTRATION']) async {
    final response = await http.post(
      Uri.parse('$baseUrl/auth/verify-otp'),
      headers: _getHeaders(),
      body: json.encode({'target': target, 'otp_code': otpCode, 'otp_type': otpType}),
    );
    final resBody = json.decode(response.body);
    if (response.statusCode != 200) {
      throw Exception(resBody['error']?['message'] ?? 'Invalid OTP code');
    }
    setTokens(resBody['access_token'], resBody['refresh_token']);
    return resBody;
  }

  Future<Map<String, dynamic>> login({
    required String identifier,
    String? password,
    String? otpCode,
    String loginType = 'password',
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/auth/login'),
      headers: _getHeaders(),
      body: json.encode({
        'identifier': identifier,
        'password': password,
        'otp_code': otpCode,
        'login_type': loginType,
      }),
    );
    final resBody = json.decode(response.body);
    if (response.statusCode != 200) {
      throw Exception(resBody['error']?['message'] ?? 'Login failed');
    }
    setTokens(resBody['access_token'], resBody['refresh_token']);
    return resBody;
  }

  Future<Map<String, dynamic>> getMe() async {
    final response = await http.get(
      Uri.parse('$baseUrl/registration/me'),
      headers: _getHeaders(),
    );
    if (response.statusCode != 200) {
      throw Exception('Failed to fetch profile details');
    }
    return json.decode(response.body);
  }

  Future<Map<String, dynamic>> saveDraft(int step, Map<String, dynamic> draftData) async {
    final response = await http.put(
      Uri.parse('$baseUrl/profile/draft'),
      headers: _getHeaders(),
      body: json.encode({'current_step': step, 'draft_data': draftData}),
    );
    if (response.statusCode != 200) {
      throw Exception('Failed to save draft');
    }
    return json.decode(response.body);
  }

  Future<Map<String, dynamic>> submitRegistration() async {
    final response = await http.post(
      Uri.parse('$baseUrl/registration/submit'),
      headers: _getHeaders(),
      body: json.encode({'confirmed': true}),
    );
    if (response.statusCode != 200) {
      throw Exception('Failed to submit profile');
    }
    return json.decode(response.body);
  }

  // ------------------ PHOTOS & VERIFICATION ------------------
  Future<Map<String, dynamic>> getMyPhotos() async {
    final response = await http.get(
      Uri.parse('$baseUrl/photos/my'),
      headers: _getHeaders(),
    );
    if (response.statusCode != 200) throw Exception('Failed to fetch photos');
    return json.decode(response.body);
  }

  Future<Map<String, dynamic>> getVerificationStatus() async {
    final response = await http.get(
      Uri.parse('$baseUrl/verification/status'),
      headers: _getHeaders(),
    );
    if (response.statusCode != 200) throw Exception('Failed to fetch verification status');
    return json.decode(response.body);
  }

  // ------------------ DISCOVERY & CHAT ------------------
  Future<Map<String, dynamic>> searchProfiles([Map<String, dynamic>? params]) async {
    final uri = Uri.parse('$baseUrl/discovery/search');
    final response = await http.get(uri, headers: _getHeaders());
    if (response.statusCode != 200) throw Exception('Failed to load profiles');
    return json.decode(response.body);
  }

  Future<Map<String, dynamic>> sendInterest(int targetUserId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/interests/send'),
      headers: _getHeaders(),
      body: json.encode({'target_user_id': targetUserId}),
    );
    if (response.statusCode != 200) throw Exception('Failed to send interest');
    return json.decode(response.body);
  }

  Future<Map<String, dynamic>> getInterests(String tab) async {
    final response = await http.get(
      Uri.parse('$baseUrl/interests?tab=$tab'),
      headers: _getHeaders(),
    );
    if (response.statusCode != 200) throw Exception('Failed to load interests');
    return json.decode(response.body);
  }

  Future<List<dynamic>> getChatHistory(int otherUserId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/chat/$otherUserId'),
      headers: _getHeaders(),
    );
    if (response.statusCode != 200) throw Exception('Failed to get chat messages');
    return (json.decode(response.body))['messages'] as List<dynamic>;
  }

  Future<Map<String, dynamic>> sendMessage(int otherUserId, String text) async {
    final response = await http.post(
      Uri.parse('$baseUrl/chat/$otherUserId'),
      headers: _getHeaders(),
      body: json.encode({'message_text': text}),
    );
    if (response.statusCode != 200) throw Exception('Failed to send message');
    return json.decode(response.body);
  }
}

File: web/lib/api-client.ts
import {
  CandidateCard,
  ChatMessageItem,
  HealthCheckResponse,
  InterestItem,
  ProfilePhotoItem,
  SubscriptionPlanItem,
  VerificationStatusResponse,
} from '../types';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

class ApiClient {
  private getHeaders(isFormData = false): HeadersInit {
    const headers: Record<string, string> = {};
    if (!isFormData) {
      headers['Content-Type'] = 'application/json';
    }
    if (typeof window !== 'undefined') {
      const token = localStorage.getItem('access_token');
      if (token) {
        headers['Authorization'] = `Bearer ${token}`;
      }
    }
    return headers;
  }

  async getHealth(): Promise<HealthCheckResponse> {
    const res = await fetch(`${API_BASE_URL}/health`);
    if (!res.ok) throw new Error('Health check failed');
    return res.json();
  }

  // ------------------ AUTH & REGISTRATION ------------------
  async register(data: any): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/auth/register`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(data),
    });
    const resData = await res.json();
    if (!res.ok) throw new Error(resData.error?.message || 'Registration failed');
    return resData;
  }

  async sendOtp(target: string, otpType = 'REGISTRATION'): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/auth/send-otp`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ target, otp_type: otpType }),
    });
    const resData = await res.json();
    if (!res.ok) throw new Error(resData.error?.message || 'Failed to send OTP');
    return resData;
  }

  async verifyOtp(target: string, otpCode: string, otpType = 'REGISTRATION'): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/auth/verify-otp`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ target, otp_code: otpCode, otp_type: otpType }),
    });
    const resData = await res.json();
    if (!res.ok) throw new Error(resData.error?.message || 'Verification failed');
    if (typeof window !== 'undefined' && resData.access_token) {
      localStorage.setItem('access_token', resData.access_token);
      if (resData.refresh_token) localStorage.setItem('refresh_token', resData.refresh_token);
    }
    return resData;
  }

  async login(payload: any): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/auth/login`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });
    const resData = await res.json();
    if (!res.ok) throw new Error(resData.error?.message || 'Login failed');
    if (typeof window !== 'undefined' && resData.access_token) {
      localStorage.setItem('access_token', resData.access_token);
      if (resData.refresh_token) localStorage.setItem('refresh_token', resData.refresh_token);
      if (resData.role) localStorage.setItem('user_role', resData.role);
    }
    return resData;
  }

  async logout(): Promise<void> {
    try {
      if (typeof window !== 'undefined') {
        const refresh = localStorage.getItem('refresh_token');
        if (refresh) {
          await fetch(`${API_BASE_URL}/auth/logout`, {
            method: 'POST',
            headers: this.getHeaders(),
            body: JSON.stringify({ refresh_token: refresh }),
          });
        }
      }
    } catch {
      // Ignored
    } finally {
      if (typeof window !== 'undefined') {
        localStorage.removeItem('access_token');
        localStorage.removeItem('refresh_token');
      }
    }
  }

  async getMe(): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/registration/me`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to fetch user state');
    return res.json();
  }

  async getRegistrationMe(): Promise<any> {
    return this.getMe();
  }

  async getDraft(): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/profile/draft`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to fetch draft');
    return res.json();
  }

  async saveDraft(step: number, draftData: Record<string, any>): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/profile/draft`, {
      method: 'PUT',
      headers: this.getHeaders(),
      body: JSON.stringify({ current_step: step, draft_data: draftData }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to save draft');
    return data;
  }

  async submitProfile(confirmed = true): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/registration/submit`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ confirmed }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to submit profile');
    return data;
  }

  async submitRegistration(confirmed = true): Promise<any> {
    return this.submitProfile(confirmed);
  }

  // ------------------ PHOTOS ------------------
  async uploadPhoto(file: File, isPrimary = false): Promise<ProfilePhotoItem> {
    const formData = new FormData();
    formData.append('file', file);
    const res = await fetch(`${API_BASE_URL}/photos/upload?is_primary=${isPrimary}`, {
      method: 'POST',
      headers: this.getHeaders(true),
      body: formData,
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Photo upload failed');
    return data;
  }

  async getMyPhotos(): Promise<{ photos: ProfilePhotoItem[]; count: number; has_min_5: boolean }> {
    const res = await fetch(`${API_BASE_URL}/photos/my`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to fetch photos');
    return res.json();
  }

  async setPrimaryPhoto(photoId: number): Promise<void> {
    const res = await fetch(`${API_BASE_URL}/photos/${photoId}/primary`, {
      method: 'PUT',
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to set primary photo');
  }

  async deletePhoto(photoId: number): Promise<void> {
    const res = await fetch(`${API_BASE_URL}/photos/${photoId}`, {
      method: 'DELETE',
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to delete photo');
  }

  // ------------------ VERIFICATION ------------------
  async getVerificationStatus(): Promise<VerificationStatusResponse> {
    const res = await fetch(`${API_BASE_URL}/verification/status`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to get verification status');
    return res.json();
  }

  // ------------------ DISCOVERY ------------------
  async searchProfiles(params: Record<string, any> = {}): Promise<{ total: number; profiles: CandidateCard[] }> {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, val]) => {
      if (val !== undefined && val !== null && val !== '') {
        searchParams.append(key, String(val));
      }
    });
    const res = await fetch(`${API_BASE_URL}/discovery/search?${searchParams.toString()}`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to search profiles');
    return res.json();
  }

  async getCandidateProfile(id: number): Promise<CandidateCard & Record<string, any>> {
    const res = await fetch(`${API_BASE_URL}/discovery/profiles/${id}`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to fetch candidate profile');
    return res.json();
  }

  // ------------------ INTERESTS ------------------
  async sendInterest(targetUserId: number, message?: string): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/interests/send`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ target_user_id: targetUserId, message }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to send interest');
    return data;
  }

  async respondInterest(interestId: number, accept: boolean): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/interests/${interestId}/respond`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ accept }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to respond to interest');
    return data;
  }

  async getInterests(tab: 'received' | 'sent' | 'matches'): Promise<{ items: InterestItem[]; count: number }> {
    const res = await fetch(`${API_BASE_URL}/interests?tab=${tab}`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to get interests');
    return res.json();
  }

  // ------------------ CHAT ------------------
  async getConversations(): Promise<any[]> {
    const res = await fetch(`${API_BASE_URL}/chat/conversations`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to fetch conversations');
    const data = await res.json();
    return data.conversations;
  }

  async getChatHistory(otherUserId: number): Promise<ChatMessageItem[]> {
    const res = await fetch(`${API_BASE_URL}/chat/${otherUserId}`, {
      headers: this.getHeaders(),
    });
    if (!res.ok) throw new Error('Failed to get chat messages');
    const data = await res.json();
    return data.messages;
  }

  async sendMessage(otherUserId: number, text: string): Promise<ChatMessageItem> {
    const res = await fetch(`${API_BASE_URL}/chat/${otherUserId}`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ message_text: text }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to send message');
    return data;
  }

  // ------------------ SUBSCRIPTIONS & PAYMENTS ------------------
  async getPlans(): Promise<SubscriptionPlanItem[]> {
    const res = await fetch(`${API_BASE_URL}/subscriptions/plans`);
    if (!res.ok) throw new Error('Failed to fetch plans');
    const data = await res.json();
    return data.plans;
  }

  async createSubscriptionOrder(planId: number): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/subscriptions/create-order`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ plan_id: planId }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to create order');
    return data;
  }

  async verifyPayment(orderId: string, paymentId: string, signature: string): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/subscriptions/verify-payment`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({
        order_id: orderId,
        razorpay_order_id: orderId,
        gateway_payment_id: paymentId,
        razorpay_payment_id: paymentId,
        gateway_signature: signature,
        razorpay_signature: signature,
      }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || data.detail || 'Payment verification failed');
    return data;
  }

  // ------------------ CONTACT REVEAL ------------------
  async requestContactReveal(targetUserId: number): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/contact-reveal/request/${targetUserId}`, {
      method: 'POST',
      headers: this.getHeaders(),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Failed to request contact reveal');
    return data;
  }

  async getContactDetails(targetUserId: number): Promise<any> {
    const res = await fetch(`${API_BASE_URL}/contact-reveal/${targetUserId}`, {
      headers: this.getHeaders(),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error?.message || 'Contact reveal not completed or paid');
    return data;
  }
}

export const apiClient = new ApiClient();
